Adds Tado Child Lock support

proohit opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds support for tado child lock (https://support.tado.com/en/articles/4849684-what-is-child-lock).

Also bumps python-tado to 0.18.6 which introduced this feature api wise (https://github.com/wmalgadey/PyTado/releases/tag/0.18.6).

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Feature request: https://community.home-assistant.io/t/child-lock-on-tado-radiator-valves/284227
## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository] https://github.com/home-assistant/home-assistant.io/pull/37120

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

proohit committed:
feat(tado): adds child_lock switch WIP

proohit committed:
feat(tado): adds child lock to zone instead

proohit commented:
Hey, 

this is my first contribution and I'm not yet warm with all the requirements. 

The core functionality should be implemented, but there might be cases that I'm not aware of. Also, if I have some more time, I will look into testing and docs. If anyone is interested in helping out or advising, feel free :)

erwindouna commented:
Thanks for submitting your first PR! The Tado Connector is currently under a revamp, to be upgraded to a Data Update Coordinator: https://github.com/home-assistant/core/pull/134175. This PR is also completed. This will heavily impact the main structure of communicating with Tado.

How have you tested the child lock (other than going a version in the upstream)? Since this isn't released yet in the current PyTado library. I have recently added this [particular code](https://github.com/wmalgadey/PyTado/blob/a36470e80cb3426573db9d6f4a98bc38570e4141/PyTado/interface/api/my_tado.py#L403) myself in the PyTado library. 😄 
It will be released soon and when it does, I personally also want to include this in HA. It will once, the PyTado is scheduled for a new release and once the DUC PR is finalized.

proohit commented:
@erwindouna Thanks for your feedback. I had no other way than using the upstream^^

proohit commented:
@erwindouna Maybe this PR could assist you in that it already adds a `SwitchEntity` to a zone using the first zone device (I was heavily influenced by `climate.py` 😄  ), which you can build upon.

erwindouna commented:
It's great to see other people also wanting to contribute! If you'd like I can keep this open and once PyTado and DUC are ready, you want to integrate the child lock? :)

proohit commented:
I looked into the DUC PR and I think the changes are a very welcome alteration in how Tado data is kept stored.

If you want, you can add it yourself. If this PR helps you, you can leave it open for reference, although I think you're perfectly capable of implementing it much better :D. I can help though if needed, as I am interested in that feature being added. Will leave the decision up to you.

Thanks for enhancing the integration!

proohit committed:
refactor(tado): reverts back to default Tado interface

proohit committed:
Merge branch 'dev' into feature/tado-child-lock-support

proohit committed:
feat(tado): bumps python-tado to 0.18.6

proohit committed:
feat(tado): uses correct tado class variable

proohit committed:
feat(tado): requests coordinator update after switch

proohit commented:
@erwindouna happy to report that with 0.18.6 I have now adapted the switch to DUC and everything is working smoothly!

erwindouna commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,108 @@
+        device_info,
+    ) -> None:
+        """Initialize the Tado child lock switch entity."""
+        self._tado = coordinator
```
This can be removed. `self.coordinator` is available.

erwindouna commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,108 @@
+        """Return true if the entity is on."""
+        return self._state
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
```
Why the `**kwargs`?

proohit replied:
This ultimately comes from the `W7431:hass-argument-type` lint rule which enforces the signature to match the one from `ToggleEntity`, which we are extending. See [homeassistant/helpers/entity.py#ToggleEntity#async_turn_on]

erwindouna commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,108 @@
+    @callback
+    def _async_update_callback(self) -> None:
+        """Handle update callbacks."""
+        _LOGGER.info("Update device data")
```
If we want to inform users, inform them a bit more detailed what we're updating. :)

erwindouna reviewed (changes requested):
Great additions, thanks @proohit! I have left the first initial comments. Also, please update the test cases, so that the child lock is also tested.

proohit committed:
refactor(tado): removes unnecessary coordinator copy

according to code review

proohit committed:
test(tado): adds tests for child lock switch

proohit committed:
test(tado): removes excess patch

proohit commented:
@erwindouna Added all the requested changes. I'm not fully satisfied with the tests, because I wanted to patch the updated state gotten from `get_device_info` but unfortunately in the tests this resolved the switch state to `unavailable`. I've tried with a custom fixture and a custom dict 

```python
patch("homeassistant.components.tado.PyTado.interface.api.Tado.get_device_info",
           return_value={"shortSerialNo": "WR4", "childLockEnabled": True},
)
```

but both had the same result.

Only thing I tested now is if the child lock api endpoint is called with the correct boolean.

proohit commented:
Also I think by updating some fixtures, I broke some snapshot tests. Do you know how to update them?

erwindouna commented:
> Also I think by updating some fixtures, I broke some snapshot tests. Do you know how to update them? 

By using `--snapshot-update`. Be careful though, you need to verify that the new snapshot really provides the new reality. Updating it for the sake of updating is not adding any value of thorough test.

proohit committed:
fix(tado): undo name change in climate

proohit commented:
Hey @erwindouna turned out the failing tests weren't from snapshots but because I changed something in climate.py that broke the tests. Undid that now, and the tests pass.

Please have another look.

erwindouna commented on homeassistant/components/tado/coordinator.py:
```diff
@@ -339,6 +339,19 @@ async def set_meter_reading(self, reading: int) -> dict[str, Any]:
 
+    async def set_child_lock(self, device_id: str, enabled: bool) -> None:
+        """Set child lock of device."""
+        if self._tado is None:
```
This doesn't seem needed. The meter reading is using a different service than the generic API to control the devices and it is setup automatically. Check other functions that do updates.

erwindouna commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,106 @@
+        """Initialize the Tado child lock switch entity."""
+        super().__init__(zone_name, coordinator.home_id, zone_id, coordinator)
+
+        self._device_info = device_info
```
You typed the bool, which is good. Please also apply typing for the other properties. :)

proohit replied:
I'm not too familiar with the typed that comes from the device_info endpoint, but I adapted it to the typed that was used in the coordinator before (which is quite generic) :)

erwindouna commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,106 @@
+async def async_setup_entry(
+    hass: HomeAssistant, entry: TadoConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Tado climate platform."""
```
Are you setting up the climate platform? ;)

erwindouna commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,106 @@
+async def _generate_entities(
+    tado: TadoDataUpdateCoordinator,
+) -> list[TadoChildLockSwitchEntity]:
+    """Create all climate entities."""
```
Climate entities? ;)

erwindouna commented on tests/components/tado/test_switch.py:
```diff
@@ -0,0 +1,65 @@
+
+    await async_init_integration(hass)
+
+    with (
```
```suggestion
    with patch(
            "homeassistant.components.tado.PyTado.interface.api.Tado.set_child_lock"
        ) as mock_set_state:
```

erwindouna commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,106 @@
+    entities: list[TadoChildLockSwitchEntity] = []
+    for zone in tado.zones:
+        zoneChildLockSupported = (
+            zone["type"] in [TYPE_HEATING]
```
Why is the `TYPE_HEATING` needed to check if the child lock is enabled?

proohit replied:
I don't know if it's relevant. I only have type heating devices here to test^^ But I think if `childLockEnabled` is present as a property, we can assume that it can be toggled too.

erwindouna commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,106 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+TRANSLATION_KEY = "child_lock"
```
Try to put this in the `const.py` file to keep contants all together.

proohit commented on homeassistant/components/tado/manifest.json:
```diff
@@ -14,5 +14,5 @@
   },
   "iot_class": "cloud_polling",
   "loggers": ["PyTado"],
-  "requirements": ["python-tado==0.18.5"]
```
Thanks for the feedback. I fear this is a direct necessity for this pr, otherwise the changes would not work. Do you have a suggestion on how to proceed here?

joostlek replied:
First open the PR to bump and then we can proceed with this PR

erwindouna replied:
@proohit you want me take it over?

erwindouna commented:
PR to bump PyTado made, to speed up progress on this: https://github.com/home-assistant/core/pull/137655

proohit committed:
chore(tado): undoes version bump

in favor of https://github.com/home-assistant/core/pull/137655

proohit committed:
Merge branch 'dev' into feature/tado-child-lock-support

proohit committed:
chore(tado): applies suggestions from code review

- fixes typos in comments
- refactors constant into const.py
- adds typings

abmantis commented on homeassistant/components/tado/const.py:
```diff
@@ -237,3 +237,5 @@
 TADO_VERTICAL_SWING_SETTING = "verticalSwing"
 TADO_HORIZONTAL_SWING_SETTING = "horizontalSwing"
+
+TADO_CHILD_LOCK_TRANSLATION_KEY = "child_lock"
```
no need to have a const for this, you can use the string directly

abmantis commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,99 @@
+class TadoChildLockSwitchEntity(TadoZoneEntity, SwitchEntity):
+    """Representation of a Tado child lock switch entity."""
+
+    _attr_unique_id: str | None = None
+
```
```suggestion
```
Is this needed?

abmantis commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,99 @@
+        self._device_info = device_info
+        self._device_id = self._device_info["shortSerialNo"]
+        self._state: bool | None = None
+        self._attr_unique_id = f"{zone_name}-child-lock"
```
unique id needs to be stable. I assume that `zone_name` can be changed by the user and is not stable, right?
the `climate` platform uses `self._attr_unique_id = f"{zone_type} {zone_id} {coordinator.home_id}"` so I suggest using a similar id here

proohit replied:
No it wasn't stable and I didn't understand what it did, until I changed the zone name and realized. Thanks for the warning.

I set it according to zone id and home id now, as I don't think the zone type is necessary here. If the underyling device would change, this would still use the same switch entity. Only when the zone id changes (tado wise) this would lead to "duplicate" entities. But this applies to other entities in home assistant tado platform as well

abmantis commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,99 @@
+        self._attr_unique_id = f"{zone_name}-child-lock"
+
+    @property
+    def translation_key(self) -> str:
```
you can set `_attr_translation_key = "child_lock"` as a class attribute

abmantis commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,99 @@
+        return TADO_CHILD_LOCK_TRANSLATION_KEY
+
+    @property
+    def is_on(self) -> bool | None:
```
we can remove this method and set `_attr_is_on` directly instead of `_state`

abmantis commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,99 @@
+        try:
+            self._device_info = self.coordinator.data["device"][self._device_id]
+        except KeyError:
+            return
```
should we log something here, or is this an expected situation?

proohit committed:
refactor(tado): applies code review suggestions

- sets truly unique id
- removes unnecessary property accessors
- adds error logging

abmantis commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,93 @@
+                self._device_id,
+                self.zone_name,
+            )
+        self._attr_is_on = self._device_info.get("childLockEnabled", False) is True
```
```suggestion
        else:
            self._attr_is_on = self._device_info.get("childLockEnabled", False) is True
```
this should be on the else branch, right?

proohit replied:
I think it should update the state nevertheless.

abmantis replied:
But is it ok to use `self._device_info` without getting a new instance? If so, do we need to always fetch a new instance?

proohit replied:
The fetching doesn't happen in this place. As far as I understand the data coordinator notifies subscribed entities if new data is available. 

I don't know if we really need to try/except but in case the device is deleted, I'm not sure if this deletion would reflect correctly for this entity. So better try to find the device info if available and nevertheless update the state from the cached (or updated) device info. 

It's really just some kind of "finally" and maybe it's better to put it in a "finally" block^^

proohit replied:
Done! And thanks for the heads up.

abmantis commented on tests/components/tado/test_switch.py:
```diff
@@ -0,0 +1,61 @@
+
+from .util import async_init_integration
+
+child_lock_switch_entity = "switch.baseboard_heater_child_lock"
```
we should use ALL_CAPS for consts

abmantis commented on tests/components/tado/test_switch.py:
```diff
@@ -0,0 +1,61 @@
+    assert mock_set_state.call_args[0][1] is True
+
+
+async def test_set_child_lock_off(hass: HomeAssistant) -> None:
```
we could merge this test and the one above into 1 single test by using `@pytest.mark.parametrize`

abmantis commented:
Just a few small suggestions and otherwise LGTM!

proohit committed:
refactor(tado): simplifies tests

proohit committed:
refactor(tado): adds else clause for self update

joostlek commented on homeassistant/components/tado/coordinator.py:
```diff
@@ -339,6 +339,17 @@ async def set_meter_reading(self, reading: int) -> dict[str, Any]:
+                enabled,
+            )
+        except RequestException as exc:
+            raise UpdateFailed(f"Error setting Tado child lock: {exc}") from exc
```
This is not an `UpdateFailed`, this rather is a `HomeAssistantError` as this is run in an action context

proohit replied:
Why is that? It updates the state remotely and also updates HA state afterwards. Plus, all the other methods in this api wrapper use this expection. Curious on why all the methods are fine using `UpdateFailed` but this particular one is not :)

joostlek replied:
Everything used in an updating context `_async_update_data` should raise `UpdateFailed` when not working, but everything in an action context should raise `HomeAssistantError` if it is something happened during execution and `ServiceValidationError` if it was a user mistake

proohit replied:
So basically almost all the methods defined in this coordinator are subject to change?

joostlek commented on homeassistant/components/tado/strings.json:
```diff
@@ -64,6 +64,11 @@
         }
       }
     },
+    "switch": {
+      "child_lock": {
```
can we add an icon for this as well?

joostlek commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,94 @@
+            self._attr_is_on = self._device_info.get("childLockEnabled", False) is True
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: TadoConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Tado switch platform."""
+
+    tado = entry.runtime_data.coordinator
+    entities: list[TadoChildLockSwitchEntity] = await _generate_entities(tado)
+
+    async_add_entities(entities, True)
+
+
+async def _generate_entities(
+    tado: TadoDataUpdateCoordinator,
+) -> list[TadoChildLockSwitchEntity]:
+    """Create all switch entities."""
+    entities: list[TadoChildLockSwitchEntity] = []
+    for zone in tado.zones:
+        zoneChildLockSupported = (
+            len(zone["devices"]) > 0 and "childLockEnabled" in zone["devices"][0]
+        )
+
+        if not zoneChildLockSupported:
+            continue
+
+        entities.append(
+            TadoChildLockSwitchEntity(
+                tado, zone["name"], zone["id"], zone["devices"][0]
+            )
+        )
+    return entities
```
These can be merged

joostlek commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,94 @@
+            self._attr_is_on = self._device_info.get("childLockEnabled", False) is True
+
+
+async def async_setup_entry(
```
Please have this before the class definition

joostlek commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,94 @@
+        self._device_info = device_info
+        self._device_id = self._device_info["shortSerialNo"]
+        self._attr_unique_id = f"{zone_id} {coordinator.home_id} child-lock"
+        self._attr_translation_key = "child_lock"
```
can be set outside of the constructor

joostlek commented on homeassistant/components/tado/switch.py:
```diff
@@ -0,0 +1,94 @@
+
+        self._device_info = device_info
+        self._device_id = self._device_info["shortSerialNo"]
+        self._attr_unique_id = f"{zone_id} {coordinator.home_id} child-lock"
```
what do we use as unique identifier seperator in the other platforms? Also a space?

proohit replied:
From `climate.py`:

```python
self._attr_unique_id = f"{zone_type} {zone_id} {coordinator.home_id}"
```

proohit replied:
From `water_heater.py`

```python
self._attr_unique_id = f"{zone_id} {coordinator.home_id}"
```

proohit committed:
refactor(tado): applies suggested changes from code review

- added icon for lock states
- refactored switch code order
- merged methods

proohit committed:
refactor(tado): uses `HomeAssistantError` instead of `UpdateFailed` for child lock

abmantis reviewed (approved):
Thanks!

abmantis commented:
@proohit can you fix the lint issues?

proohit commented:
```
Warning: homeassistant/components/tado/switch.py:17:49: W7431: Argument 3 should be of type AddConfigEntryEntitiesCallback in async_setup_entry (hass-argument-type)
Warning: homeassistant/components/tado/switch.py:17:49: W7431: Argument 3 should be of type AddConfigEntryEntitiesCallback in async_setup_entry (hass-argument-type)
```

@abmantis unfortunately, I do not know how to fix the issues. For one, I cannot reproduce them locally:

```
(ha-venv) vscode ➜ /workspaces/core (feature/tado-child-lock-support) $ pylint homeassistant/components/tado
(ha-venv) vscode ➜ /workspaces/core (feature/tado-child-lock-support) $ pylint tests/components/tado        
(ha-venv) vscode ➜ /workspaces/core (feature/tado-child-lock-support) $ 
```

And secondly, I could not find any type/class with the identifier `AddConfigEntryEntitiesCallback `. I searched in all the repository and there was nothing found with that name. Therefore, if I change the `async_setup_entry` implementation of the switch, I cannot import this.

The documentation does not use any types for `async_setup_entry` at all (see [documentation on config entries](https://developers.home-assistant.io/docs/config_entries_index/#for-platforms)).

If I remove the type declarations altogether, pylint pops up warnings:

```
(ha-venv) vscode ➜ /workspaces/core (feature/tado-child-lock-support) $ pylint homeassistant/components/tado
************* Module homeassistant.components.tado.switch
homeassistant/components/tado/switch.py:14:28: W7431: Argument hass should be of type ['HomeAssistant', 'HomeAssistant | None'] in async_setup_entry (hass-argument-type)
homeassistant/components/tado/switch.py:14:34: W7431: Argument 2 should be of type ConfigEntry in async_setup_entry (hass-argument-type)
homeassistant/components/tado/switch.py:14:41: W7431: Argument 3 should be of type AddEntitiesCallback in async_setup_entry (hass-argument-type)
homeassistant/components/tado/switch.py:14:28: W7431: Argument hass should be of type ['HomeAssistant', 'HomeAssistant | None'] in async_setup_entry (hass-argument-type)
homeassistant/components/tado/switch.py:14:34: W7431: Argument 2 should be of type ConfigEntry in async_setup_entry (hass-argument-type)
homeassistant/components/tado/switch.py:14:41: W7431: Argument 3 should be of type AddEntitiesCallback in async_setup_entry (hass-argument-type)
```

In conclusion, I don't know how to fix that as it seems like a false positive for me.

abmantis commented:
@proohit you need to merge dev into your branch so it gets the latest changes (including `AddConfigEntryEntitiesCallback`).

proohit committed:
Merge branch 'dev' into feature/tado-child-lock-support

proohit committed:
chore(tado): fixes linting errors

abmantis reviewed (approved):
Thank you!
